File: batch/cmd/crawl.go
package cmd

import (
	"errors"
	"net/http"
	"strings"
	"time"

	"github.com/fjnkt98/atcodersearch-batch/crawl"
	"github.com/fjnkt98/atcodersearch-batch/pkg/atcoder"
	"github.com/fjnkt98/atcodersearch-batch/repository"
	"github.com/urfave/cli/v2"
)

var ErrEmptyAuth = errors.New("atcoder-username or atcoder-password is empty")

func NewCrawlCmd() *cli.Command {
	crawlProblemCmd := &cli.Command{
		Name: "problem",
		Flags: []cli.Flag{
			&cli.DurationFlag{
				Name:    "duration",
				Aliases: []string{"d"},
				Value:   1000 * time.Millisecond,
			},
			&cli.BoolFlag{
				Name:    "all",
				Aliases: []string{"a"},
				Value:   false,
			},
		},
		Action: func(ctx *cli.Context) error {
			problemsClient := atcoder.NewAtCoderProblemsClientWithHTTPClient(
				&http.Client{
					Timeout: 5 * time.Second,
				},
			)
			atcoderClient, err := atcoder.NewAtCoderClientWithHTTPClient(
				&http.Client{
					Timeout: 5 * time.Second,
				},
			)
			if err != nil {
				return err
			}

			pool, err := repository.NewPool(ctx.Context, ctx.String("database-url"))
			if err != nil {
				return err
			}

			if err := crawl.NewContestCrawler(problemsClient, pool).Crawl(ctx.Context); err != nil {
				return err
			}
			if err := crawl.NewDifficultyCrawler(problemsClient, pool).Crawl(ctx.Context); err != nil {
				return err
			}
			if err := crawl.NewProblemCrawler(
				problemsClient,
				atcoderClient,
				pool,
				ctx.Duration("duration"),
				ctx.Bool("all"),
			).Crawl(ctx.Context); err != nil {
				return err
			}

			return nil
		},
	}
	crawlUserCmd := &cli.Command{
		Name: "user",
		Flags: []cli.Flag{
			&cli.DurationFlag{
				Name:    "duration",
				Aliases: []string{"d"},
				Value:   1000 * time.Millisecond,
			},
		},
		Action: func(ctx *cli.Context) error {
			client, err := atcoder.NewAtCoderClientWithHTTPClient(
				&http.Client{
					Timeout: 5 * time.Second,
				},
			)
			if err != nil {
				return err
			}

			pool, err := repository.NewPool(ctx.Context, ctx.String("database-url"))
			if err != nil {
				return err
			}

			if err := crawl.NewUserCrawler(client, pool, ctx.Duration("duration")).Crawl(ctx.Context); err != nil {
				return err
			}

			return nil
		},
	}
	crawlSubmissionCmd := &cli.Command{
		Name: "submission",
		Flags: []cli.Flag{
			&cli.DurationFlag{
				Name:    "duration",
				Aliases: []string{"d"},
				Value:   1000 * time.Millisecond,
			},
			&cli.IntFlag{
				Name:    "retry",
				Aliases: []string{"r"},
				Value:   0,
			},
			&cli.StringFlag{
				Name: "target",
			},
			&cli.StringFlag{
				Name:    "atcoder-username",
				EnvVars: []string{"ATCODER_USERNAME"},
			},
			&cli.StringFlag{
				Name:    "atcoder-password",
				EnvVars: []string{"ATCODER_PASSWORD"},
			},
		},
		Action: func(ctx *cli.Context) error {
			client, err := atcoder.NewAtCoderClientWithHTTPClient(
				&http.Client{
					Timeout: 5 * time.Second,
				},
			)
			if err != nil {
				return err
			}

			username := ctx.String("atcoder-username")
			password := ctx.String("atcoder-password")
			if username == "" || password == "" {
				return ErrEmptyAuth
			}

			if err := client.Login(ctx.Context, username, password); err != nil {
				return err
			}

			pool, err := repository.NewPool(ctx.Context, ctx.String("database-url"))
			if err != nil {
				return err
			}

			var targets []string
			if target := ctx.String("target"); target != "" {
				targets = strings.Split(target, ",")
			}

			crawler := crawl.NewSubmissionCrawler(
				client,
				pool,
				ctx.Duration("duration"),
				ctx.Int("retry"),
				targets,
			)
			if err := crawler.Crawl(ctx.Context); err != nil {
				return err
			}

			return nil
		},
	}

	return &cli.Command{
		Name: "crawl",
		Subcommands: []*cli.Command{
			crawlProblemCmd,
			crawlUserCmd,
			crawlSubmissionCmd,
		},
	}
}

File: batch/src/cmd/crawl.rs
use std::time::Duration;

use crate::{
    atcoder::{AtCoderClient, AtCoderProblemsClient},
    crawl::{contest, difficulty, problem, submission, user},
};
use anyhow::Context;
use clap::Subcommand;
use itertools::Itertools;
use sqlx::postgres::PgPoolOptions;

use super::CommonArgs;

#[derive(Subcommand)]
pub(crate) enum CrawlCommand {
    Problem {
        #[arg(long, default_value = "1s")]
        duration: humantime::Duration,
        #[arg(short, long, help = "if true, crawl all problems.")]
        all: bool,
    },
    User {
        #[arg(long, default_value_t = 1)]
        duration: u64,
    },
    Submission {
        #[arg(long, default_value = "1s")]
        duration: humantime::Duration,
        #[arg(
            long,
            default_value_t = 0,
            help = "number of retries allowed when crawling failed. zero means no retry."
        )]
        retry: i64,
        #[arg(
            long,
            default_value = "",
            help = "comma separated list of target contest id. if not specified, crawl all contests."
        )]
        target: String,
        #[arg(long, env, hide_env_values = true)]
        atcoder_username: String,
        #[arg(long, env, hide_env_values = true)]
        atcoder_password: String,
    },
}

impl CrawlCommand {
    pub async fn exec(&self, args: &CommonArgs) -> anyhow::Result<()> {
        let problems_client = AtCoderProblemsClient::new()?;
        let atcoder_client = AtCoderClient::new()?;
        let pool = PgPoolOptions::new()
            .connect(&args.database_url)
            .await
            .with_context(|| "connect to database")?;

        match self {
            Self::Problem { duration, all } => {
                contest::crawl_contests(&problems_client, &pool)
                    .await
                    .with_context(|| "crawl contests")?;

                difficulty::crawl_difficulties(&problems_client, &pool)
                    .await
                    .with_context(|| "crawl difficulties")?;

                problem::crawl_problems(
                    &atcoder_client,
                    &problems_client,
                    &pool,
                    *all,
                    (*duration).into(),
                )
                .await
                .with_context(|| "crawl problems")?;

                Ok(())
            }
            Self::User { duration } => {
                user::crawl_users(&atcoder_client, &pool, Duration::from_secs(*duration))
                    .await
                    .with_context(|| "crawl users")?;

                Ok(())
            }
            Self::Submission {
                duration,
                retry,
                target,
                atcoder_username,
                atcoder_password,
            } => {
                atcoder_client
                    .login(atcoder_username, atcoder_password)
                    .await
                    .with_context(|| "login to atcoder")?;

                let targets = target.split(",").map(String::from).collect_vec();
                let crawler = submission::SubmissionCrawler::new(
                    atcoder_client,
                    pool,
                    (*duration).into(),
                    *retry,
                    targets,
                );

                crawler.crawl().await?;
                Ok(())
            }
        }
    }
}
